bluemira/optimisation/_optimiser.py

```diff
@@ -0,0 +1,165 @@
+        constraint of the form $h(x) = 0$.
+
+        The constraint function must have the form `f(x) -> y`,
+        where:
+
+            * `x` is a numpy array of the optimisation parameters.
+            * `y` is a numpy array containing the values of the
+              constraint at `x`, with size $m$, where $m$ is the
+              dimensionality of the constraint.
+
+        Parameters
+        ----------
+        f_constraint: Callable[[Arg(np.ndarray, 'x')], np.ndarray]
+            The constraint function, with form as described above.
+        tolerance: np.ndarray
+            The tolerances for each optimisation parameter.
+        df_constraint: Optional[Callable[[Arg(np.ndarray, 'x')], np.ndarray]]
+            The gradient of the constraint function. This should have the
+            same form as the constraint function, however its output
+            array should have dimensions `m x n_variables` where `m` is
+            the dimensionality of the constraint.
+
+        Notes
+        -----
+        Equality constraints are only supported by algorithms:
+
+            * SLSQP
+            * COBYLA
+            * ISRES
+
+        """
+
+    @abc.abstractmethod
+    def add_ineq_constraint(
+        self,
+        f_constraint: OptimiserCallable,
+        tolerance: np.ndarray,
+        df_constraint: Optional[OptimiserCallable] = None,
+    ) -> None:
+        r"""
+        Add an inequality constrain to the optimiser.
+
+        The constraint is a vector-valued, non-linear, inequality
+        constraint of the form $f_{c}(x) \le 0$.
+
+        The constraint function should have the form `f(x) -> y`,
+        where:
+
+            * `x` is a numpy array of the optimisation parameters.
+            * `y` is a numpy array containing the values of the
+              constraint at `x`, with size $m$, where $m$ is the
+              dimensionality of the constraint.
+
+        Parameters
+        ----------
+        f_constraint: Callable[[Arg(np.ndarray, 'x')], np.ndarray]
+            The constraint function, with form as described above.
+        tolerance: Union[float, np.ndarray]
+            The tolerances for each optimisation parameter.
+        df_constraint: Optional[Callable[[Arg(np.ndarray, 'x')], np.ndarray]]
+            The gradient of the constraint function. This should have the
+            same form as the constraint function, however its output
+            array should have dimensions `m x n_variables` where `m` is
+            the dimensionality of the constraint.
+
+        Notes
+        -----
+        Inequality constraints are only supported by algorithms:
+
+            * SLSQP
+            * COBYLA
+            * ISRES
+
+        """
+
+    @abc.abstractmethod
+    def optimise(self, x0: Optional[np.ndarray] = None) -> OptimiserResult:
+        """
+        Run the optimiser.
+
+        Parameters
+        ----------
+        x0: Optional[np.ndarray]
```

Comment:
Just fyi we don't need to type in both the function definition and the docstrings anymore. The documentation build automatically will populate both if its in the function definition. 
Not really important though and in some cases we can be more descriptive in the docstring. If the type already exists in the docstring I dont think its replaced by the function definition type.
```suggestion
        x0
```